permits/api.py

```diff
@@ -0,0 +1,186 @@
+            )
+            .prefetch_related(works_object_types_prefetch)
+            .select_related("permit_request__administrative_entity")
+        )
+
+        return qs.order_by("starts_at")
+
+
+# //////////////////////////////////
+# PERMIT REQUEST ENDPOINT
+# //////////////////////////////////
+
+
+class GeomToText(GeomOutputGeoFunc):
+    function = "ST_asText"
+    geom_param_pos = (0,)
+    output_field = CharField()
+
+
+class PermitRequestViewSet(viewsets.ReadOnlyModelViewSet):
+    """
+    Permit request endpoint Usage:
+        1.- /rest/permits/?permit-request-id=1
+        2.- /rest/permits/?works-object-type=1&status=0
+        3.- /rest/permits/?geom-type=lines | points | polygons
+    """
+
+    serializer_class = serializers.PermitRequestPrintSerializer
+
+    def get_queryset(self):
+        """
+        This view should return a list of permits for which the logged user has
+        view permissions
+        """
+        user = self.request.user
+
+        if not user.is_authenticated:
+            raise APIException(services.EndpointErrors.NO_AUTH.value)
+
+        work_objects_type = self.request.query_params.get("works-object-type", None)
+        status = self.request.query_params.get("status", None)
+        geom_type = self.request.query_params.get("geom-type", None)
+        permitrequest_id = self.request.query_params.get("permit-request-id", None)
+
+        base_filter = Q()
+
+        if work_objects_type:
+            if work_objects_type.isdigit():
+                base_filter &= Q(works_object_types=work_objects_type)
+            else:
+                raise APIException(services.EndpointErrors.WOT_NOT_INT.value)
+
+        if status:
+            if status.isdigit():
+                base_filter &= Q(status=status)
+            else:
+                raise APIException(services.EndpointErrors.STATUS_NOT_INT.value)
+
+        if permitrequest_id:
+            if permitrequest_id.isdigit():
+                base_filter &= Q(pk=permitrequest_id)
+            else:
+                raise APIException(services.EndpointErrors.PR_NOT_INT.value)
+
+        works_object_types_prefetch = Prefetch(
+            "works_object_types",
+            queryset=models.WorksObjectType.objects.select_related("works_type"),
+        )
+
+        geom_qs = models.PermitRequestGeoTime.objects.only("geom")
+
+        if geom_type:
+            if geom_type not in ("lines", "points", "polygons"):
+                raise APIException(services.EndpointErrors.GEO_NOT_VALID.value)
+            geom_qs = geom_qs.annotate(geom_type=GeomToText(F("geom"),))
+            if geom_type == "lines":
+                geom_qs = geom_qs.filter(geom_type__icontains="line")
+            if geom_type == "points":
+                geom_qs = geom_qs.filter(geom_type__icontains="point")
+            if geom_type == "polygons":
+                geom_qs = geom_qs.filter(geom_type__icontains="poly")
+            base_filter &= Q(id__in=geom_qs)
```

Comment:
Je suis pas sûr de comprendre ce filtre. `base_filter` s’applique à des `PermitRequest`, donc là tu récupères les permit requests dont l’id est un des id du queryset `PermitRequestGeoTime`?

Reply:
If the user wants to filter by a specific geometry type:

1. We annotate the geometry types on the PermitRequestGeoTime queryset
2. We filter that queryset by the geometry type we want, thus having the ids of all the related permit request 
3. Finally, the base_filter will be in charge of passing to the serializer a qs with only the concerned permit requests

Note that the base_filter is the combination of all the existing filters (work_objects_type, status, geom_type, permit_request_id).

Reply:
base_filter modified, the id must be within the values of the permit_request_id, not the PermitRequestGeoTime id.